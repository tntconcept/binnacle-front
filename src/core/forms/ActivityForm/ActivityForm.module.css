.base {
  display: grid;
  grid-template-areas:
          "startTime endTime"
          "duration duration"
          "entities entities"
          "billable billable"
          "description description";
  grid-template-columns: minmax(0, 1fr) minmax(0, 1fr);
  grid-column-gap: 8px;
  grid-row-gap: 16px;
  padding: 16px;
}

.startTime {
  grid-area: startTime;
}

.endTime {
  grid-area: endTime;
}

.duration {
  grid-area: duration;
}

.entities {
  grid-area: entities;
}

.billable {
  grid-area: billable;
}

.description {
  grid-area: description;
}

.fieldset {
  composes: entities;
  border: none;
  padding: 0;
  margin: 0;
}

.legend {
  margin-bottom: 16px;
}

.entitiesContainer {
  background-color: #fbfbfc;
  display: grid;
  grid-gap: 16px;
}