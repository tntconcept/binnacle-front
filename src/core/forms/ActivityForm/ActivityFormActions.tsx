import React, {memo} from "react"
import useModal from "core/hooks/useModal"
import {useTranslation} from "react-i18next"
import Button from "core/components/Button"
import styles from "./ActivityForm.module.css"
import {IActivity} from "api/interfaces/IActivity"
import ErrorModal from "core/components/ErrorModal"

interface IActivityFormActions {
  activity: IActivity | undefined;
  onRemove: () => Promise<void>;
  onSave: () => void;
}

const ActivityFormActions: React.FC<IActivityFormActions> = memo(props => {
  const { t } = useTranslation();
  const { modalIsOpen, toggleModal, setIsOpen } = useModal();

  const handleDeleteActivity = async () => {
    await props.onRemove();
    setIsOpen(false);
  };

  return (
    <div
      className={styles.footer}
      style={{
        justifyContent: props.activity ? "space-between" : "flex-end"
      }}
    >
      {modalIsOpen && (
        <ErrorModal
          message={{
            title: t("remove_modal.title"),
            description: t("remove_modal.description")
          }}
          onClose={() => setIsOpen(false)}
          onConfirm={handleDeleteActivity}
          confirmText={t("actions.remove")}
        />
      )}
      {props.activity && (
        <Button
          data-testid="remove_activity"
          isTransparent
          onClick={toggleModal}
          type={"button"}
        >
          {t("actions.remove")}
        </Button>
      )}
      <Button data-testid="save_activity" type="button" onClick={props.onSave}>
        {t("actions.save")}
      </Button>
    </div>
  );
});

export default ActivityFormActions;
