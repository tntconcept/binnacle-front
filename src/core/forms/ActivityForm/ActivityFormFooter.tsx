import React, {memo} from "react"
import styles from "core/forms/ActivityForm/ActivityForm.module.css"
import useModal from "core/hooks/useModal"
import Modal from "core/components/Modal"
import {useTranslation} from "react-i18next"
import {deleteActivity} from "services/activitiesService"
import {useHistory} from "react-router-dom"

interface IActivityFormFooter {
  id: number | undefined;
  onRemove: () => void;
  onSave: () => void;
}

const ActivityFormFooter: React.FC<IActivityFormFooter> = memo(props => {
  const { t } = useTranslation();
  let history = useHistory();
  const { modalIsOpen, toggleIsOpen } = useModal();

  const handleDeleteActivity = async () => {
    await deleteActivity(props.id!);
    history.push("/binnacle")
  };

  return (
    <div className={styles.footer}>
      {modalIsOpen && (
        <Modal
          ariaLabel="Are you sure that you want to delete the activity?"
          onClose={toggleIsOpen}
        >
          <button data-testid="yes_modal_button" onClick={handleDeleteActivity}>
            Remove
          </button>
          <button data-testid="no_modal_button" onClick={toggleIsOpen}>
            Cancel
          </button>
        </Modal>
      )}
      {props.id && (
        <button onClick={toggleIsOpen} type={"button"}>
          {t("actions.remove")}
        </button>
      )}
      <button data-testid="save_activity" type="submit" onClick={props.onSave}>
        {t("actions.save")}
      </button>
    </div>
  );
});

export default ActivityFormFooter;
