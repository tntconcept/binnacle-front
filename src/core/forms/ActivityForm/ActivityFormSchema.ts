import * as Yup from "yup"
import i18n from "i18n"
import {isAfter, parse} from "date-fns"

export const ActivityFormSchema = Yup.object().shape({
  startTime: Yup.string().required(i18n.t("form_errors.field_required")),
  /*    .test("is-greater", "end time should be greater", function (value) {
        const {endTime} = this.parent
        return moment(value, "HH:mm").isSameOrAfter(moment(start, "HH:mm"))
      }),*/
  endTime: Yup.string()
    .required(i18n.t("form_errors.field_required"))
    .test("is-greater", i18n.t("form_errors.end_time_greater"), function (
      value
    ) {
      const {startTime} = this.parent
      const startDate = parse(startTime, "HH:mm", new Date())
      const endDate = parse(value, "HH:mm", new Date())

      // console.log(startDate, endDate);

      return isAfter(endDate, startDate)
    }),
  organization: Yup.object().required(i18n.t("form_errors.select_an_option")),
  project: Yup.object().required(i18n.t("form_errors.select_an_option")),
  role: Yup.object().required(i18n.t("form_errors.select_an_option")),
  billable: Yup.string().required(i18n.t("form_errors.field_required")),
  description: Yup.string()
    .required(i18n.t("form_errors.field_required"))
    .max(1024, i18n.t("form_errors.max_length"))
})
