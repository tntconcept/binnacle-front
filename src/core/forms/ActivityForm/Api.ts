import Cookies from 'js-cookie'
import {configureRefreshFetch, fetchJSON} from 'refresh-fetch'

const COOKIE_NAME = 'BINNACLE'

const retrieveToken = () => Cookies.get(COOKIE_NAME)
const saveToken = (token: string) => Cookies.set(COOKIE_NAME, token)
const clearToken = () => Cookies.remove(COOKIE_NAME)

const fetchJSONWithToken = <T>(url: string, options = {}) => {
  const token = retrieveToken()

  let optionsWithToken = options
  if (token != null) {
    optionsWithToken = {
      ...options,
      headers: {
        Authorization: `Bearer ${token}`
      }
    }
  }

  return fetchJSON<T>(url, optionsWithToken)
}

const login = (username: string, password: string) => {
  return fetchJSON<{token: string}>('/api/auth/login', {
    method: 'POST',
    body: JSON.stringify({
      username,
      password
    })
  })
    .then(response => {
      console.log(response)
      saveToken(response.body.token)
    })
}

const logout = () => {
  return fetchJSONWithToken('/api/auth/logout', {
    method: 'POST'
  })
    .then(() => {
      clearToken()
    })
}

const shouldRefreshToken = (error: any) =>
  error.response.status === 401 &&
  error.body.message === 'Token has expired'

const refreshToken = () => {
  return fetchJSONWithToken<{token: string}>('/api/auth/refresh-token', {
    method: 'POST'
  })
    .then(response => {
      saveToken(response.body.token)
    })
    .catch(error => {
      // Clear token and continue with the Promise catch chain
      clearToken()
      throw error
    })
}

const fetch = configureRefreshFetch({
  fetch: fetchJSONWithToken,
  shouldRefreshToken,
  refreshToken
})

export {
  fetch,
  login,
  logout
}