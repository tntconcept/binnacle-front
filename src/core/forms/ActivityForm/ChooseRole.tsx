import React from "react"
import styles from "core/forms/ActivityForm/ActivityForm.module.css"
import {useTranslation} from "react-i18next"
import {useQuery} from "react-query"
import Combobox, {IOption} from "core/components/TextField/Combobox"
import {UseComboboxState} from "downshift"
import {ORGANIZATIONS_ENDPOINT, PROJECTS_ENDPOINT} from "services/endpoints"
import {IOrganization} from "interfaces/IOrganization"
import {IProject} from "interfaces/IProject"
import {IProjectRole} from "interfaces/IProjectRole"
import {fetchClient} from "services/fetchClient"
import Index from "core/components/FieldMessage"

const fetchOrganizations = async () =>
  await fetchClient
    .url(ORGANIZATIONS_ENDPOINT)
    .get()
    .json<IOrganization[]>()

const fetchProjectsByOrganization = async ({organizationId}: any) =>
  await fetchClient
    .url(`${ORGANIZATIONS_ENDPOINT}/${organizationId}/projects`)
    .get()
    .json<IProject[]>()

const fetchRolesByProject = async ({projectId}: any) =>
  await fetchClient
    .url(`${PROJECTS_ENDPOINT}/${projectId}/roles`)
    .get()
    .json<IProjectRole[]>()

interface IChooseRole {
  formik: any;
}

const ChooseRole: React.FC<IChooseRole> = props => {
  const {t} = useTranslation()

  const organizations = useQuery<IOrganization[], {}>(
    "organizations",
    fetchOrganizations
  )

  const organizationDataExists =
    organizations.data !== null && props.formik.values.organization
  const organizationId = props.formik.values.organization
    ? props.formik.values.organization.id
    : organizationDataExists ? organizations.data![0].id : null

  const projects = useQuery<IProject[], { organizationId: number }>(
    organizationDataExists && ["projects", {organizationId: organizationId}],
    fetchProjectsByOrganization
  )

  const projectDataExists =
    projects.data !== null && props.formik.values.project
  const projectId = props.formik.values.project
    ? props.formik.values.project.id
    : projectDataExists ? projects.data![0].id : null

  const roles = useQuery<IProjectRole[], { projectId: number }>(
    projectDataExists && ["roles", {projectId: projectId}],
    fetchRolesByProject
  )

  const handleOrganizationSelect = (
    changes: Partial<UseComboboxState<IOption>>
  ) => {
    formik.setFieldValue("organization", changes.selectedItem)
  }

  const handleProjectSelect = (changes: Partial<UseComboboxState<IOption>>) => {

    if (changes.selectedItem) {
      formik.setValues({
        ...formik.values,
        // @ts-ignore
        billable: changes.selectedItem.billable ? "yes" : "no",
        project: changes.selectedItem
      })
    }
  }

  const handleProjectRoleSelect = (
    changes: Partial<UseComboboxState<IOption>>
  ) => {
    formik.setFieldValue("role", changes.selectedItem)
  }

  const {formik} = props

  const projectsDisabled = organizations.isLoading || organizations.error !== null
  const rolesDisabled = projects.isLoading || projects.error !== null

  const checkOrganizationError = () => {
    if (formik.errors.organization && formik.touched.organization) {
      return formik.errors.organization
    }

    if (organizations.error) {
      return 'Error: Failed to fetch'
    }
  }

  return (
    <div className={styles.entitiesContainer}>
      <Combobox
        label={t("activity_form.organization")}
        name="organization"
        options={organizations.data || []}
        initialSelectedItem={formik.values.organization}
        onSelect={handleOrganizationSelect}
        isLoading={organizations.isLoading}
        hasError={organizations.error}
        isDisabled={organizations.error !== null}
      >
        {checkOrganizationError() ?
          <Index text={checkOrganizationError()} isError={checkOrganizationError()}/> : null}
      </Combobox>
      <Combobox
        label={t("activity_form.project")}
        name="project"
        options={projects.data || []}
        initialSelectedItem={formik.values.project}
        onSelect={handleProjectSelect}
        isLoading={projects.isLoading}
        hasError={projects.error}
        isDisabled={projectsDisabled}
      >
        {formik.errors.project && formik.touched.project ? (
          <div>{formik.errors.project}</div>
        ) : null}
      </Combobox>
      <Combobox
        label={t("activity_form.role")}
        name="role"
        options={roles.data || []}
        initialSelectedItem={formik.values.role}
        onSelect={handleProjectRoleSelect}
        isLoading={roles.isLoading}
        hasError={roles.error}
        isDisabled={projectsDisabled || rolesDisabled}
      >
        {formik.errors.role && formik.touched.role ? (
          <div>{formik.errors.role}</div>
        ) : null}
      </Combobox>
    </div>
  )
}

export default ChooseRole
