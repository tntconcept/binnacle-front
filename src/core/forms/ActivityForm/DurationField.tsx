import React, {useState} from "react"
import {useTranslation} from "react-i18next"
import TextField from "core/components/TextField/TextField"
import {useFormikContext} from "formik"
import {lightFormat, set} from "date-fns"
import {timeToDate} from "utils/DateUtils"

const DurationField = () => {
  const { t } = useTranslation();
  const [duration, setDuration] = useState("0");
  const formik = useFormikContext()

  const handleSetDuration = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.currentTarget.value;
    const duration = Number(value) * 60;
    console.log(duration)

    // TODO activity's date or lastEndTime or currentDate as fallbackDate
    // @ts-ignore
    const endDate = timeToDate(formik.values.endTime)
    const newEndDate = set(endDate, { hours: endDate.getHours() + duration / 60, minutes: duration%60 })
    formik.setFieldValue("endTime", lightFormat(newEndDate  , "HH:mm"))

    setDuration(value)
  };

  return (
    <TextField
      label={t("activity_form.duration")}
      value={duration}
      type="number"
      onChange={handleSetDuration}
    />
  );
};

export default DurationField;
