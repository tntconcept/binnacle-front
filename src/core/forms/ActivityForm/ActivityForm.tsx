import React, {Fragment, useContext, useState} from "react"
import {Field, Formik} from "formik"
import TextField from "core/components/TextField/TextField"
import {differenceInMinutes, parse} from "date-fns"
import styles from "core/forms/ActivityForm/ActivityForm.module.css"
import {useTranslation} from "react-i18next"
import {IActivity} from "api/interfaces/IActivity"
import {IProjectRole} from "api/interfaces/IProjectRole"
import {createActivity, deleteActivity, updateActivity} from "api/ActivitiesAPI"
import ActivityFormActions from "core/forms/ActivityForm/ActivityFormActions"
import {BinnacleDataContext} from "core/contexts/BinnacleContext/BinnacleDataProvider"
import {BinnacleActions} from "core/contexts/BinnacleContext/BinnacleActions"
import FieldMessage from "core/components/FieldMessage"
import {FormikHelpers} from "formik/dist/types"
import {IProject} from "api/interfaces/IProject"
import {IOrganization} from "api/interfaces/IOrganization"
import Checkbox from "core/components/Checkbox"
import {useAutoFillHours} from "core/forms/ActivityForm/useAutoFillHours"
import {SettingsContext} from "core/contexts/SettingsContext/SettingsContext"
import DurationInput from "core/forms/ActivityForm/DurationInput"
import useRecentRoles from "core/hooks/useRecentRoles"
import {ActivityFormSchema} from "core/forms/ActivityForm/ActivityFormSchema"
import {getInitialValues} from "core/forms/ActivityForm/utils"
import DurationText from "core/forms/ActivityForm/DurationText"
import UploadImage from "core/forms/ActivityForm/UploadImage"
import ChooseRole from "core/forms/ActivityForm/ChooseRole"
import ErrorModal from "core/components/ErrorModal"
import useModal from "core/hooks/useModal"

interface IActivityForm {
  date: Date;
  activity?: IActivity;
  /** Last activity end time, fallback to settings start time value */
  lastEndTime?: Date;
  onAfterSubmit: () => void;
}

export interface ActivityFormValues {
  startTime: string;
  endTime: string;
  organization?: IOrganization;
  project?: IProject;
  role?: IProjectRole;
  billable: boolean;
  description: string;
}

const ActivityForm: React.FC<IActivityForm> = props => {
  const { t } = useTranslation();
  const {
    modalIsOpen: errorModalIsOpen,
    toggleModal: toggleErrorModal
  } = useModal();

  const { dispatch } = useContext(BinnacleDataContext);
  const { state: settingsState } = useContext(SettingsContext);
  const { startTime, endTime } = useAutoFillHours(
    settingsState.autofillHours,
    settingsState.hoursInterval,
    props.lastEndTime
  );
  const [imageBase64, setImageBase64] = useState<string | null>(null);

  const recentRoleExists = useRecentRoles(props.activity?.projectRole.id);
  const [showRecentRoles, toggleRecentRoles] = useState(
    recentRoleExists !== undefined
  );

  const handleSubmit = async (
    values: ActivityFormValues,
    formikHelpers: FormikHelpers<ActivityFormValues>
  ) => {
    if (props.activity) {
      try {
        const startDate = parse(
          values.startTime,
          "HH:mm",
          props.activity.startDate
        );
        const endTime = parse(
          values.endTime,
          "HH:mm",
          props.activity.startDate
        );
        const duration = differenceInMinutes(endTime, startDate);

        const response = await updateActivity({
          startDate: startDate,
          billable: values.billable,
          description: values.description,
          duration: duration,
          projectRoleId: values.role!.id,
          id: props.activity.id,
          hasImage: imageBase64 !== null,
          imageFile: imageBase64 !== null ? (imageBase64 as string) : undefined
        });

        dispatch(BinnacleActions.updateActivity(response));
      } catch (e) {
        toggleErrorModal();
      }
    } else {
      try {
        const startDate = parse(values.startTime, "HH:mm", props.date);
        const endTime = parse(values.endTime, "HH:mm", props.date);
        const duration = differenceInMinutes(endTime, startDate);

        const response = await createActivity({
          billable: values.billable,
          description: values.description,
          duration: duration,
          startDate: (startDate.toISOString() as unknown) as Date,
          projectRoleId: values.role!.id,
          hasImage: imageBase64 !== null,
          imageFile: imageBase64 !== null ? (imageBase64 as string) : undefined
        });

        dispatch(BinnacleActions.createActivity(response));
      } catch (e) {
        toggleErrorModal();
      }
    }

    const imputedRole = {
      id: values.role!.id,
      name: values.role!.name,
      projectName: values.project!.name,
      projectBillable: values.project!.billable,
      date: props.date
    };

    // This can be done inside create or update activity reducer, instead of here.
    if (showRecentRoles) {
      dispatch(BinnacleActions.updateLastImputedRole(imputedRole));
    } else {
      dispatch(BinnacleActions.addRecentRole(imputedRole));
    }

    props.onAfterSubmit();
  };

  const handleRemoveActivity = async () => {
    if (props.activity) {
      try {
        await deleteActivity(props.activity.id);
        dispatch(BinnacleActions.deleteActivity(props.activity));
        props.onAfterSubmit();
      } catch (e) {
        toggleErrorModal();
      }
    }
  };

  return (
    <Fragment>
      <Formik
        initialValues={getInitialValues(props.activity, recentRoleExists, {
          startTime,
          endTime
        })}
        validationSchema={ActivityFormSchema}
        onSubmit={handleSubmit}
      >
        {formik => (
          <form onSubmit={formik.handleSubmit} noValidate={true}>
            <div className={styles.base}>
              <Field
                name="startTime"
                as={TextField}
                label={t("activity_form.start_time")}
                className={styles.startTime}
                type="time"
                step="900"
              >
                <FieldMessage
                  isError={formik.errors.startTime && formik.touched.startTime}
                  errorText={formik.errors.startTime}
                />
              </Field>
              <Field
                name="endTime"
                as={TextField}
                label={t("activity_form.end_time")}
                className={styles.endTime}
                type="time"
                step="900"
              >
                <FieldMessage
                  isError={formik.errors.endTime && formik.touched.endTime}
                  errorText={formik.errors.endTime}
                />
              </Field>
              <div className={styles.duration}>
                {settingsState.showDurationInput ? (
                  <DurationInput />
                ) : (
                  <DurationText />
                )}
              </div>
              <ChooseRole
                showRecentRoles={showRecentRoles}
                toggleRecentRoles={toggleRecentRoles}
                recentRoleExists={recentRoleExists}
              />
              <Field
                name="billable"
                label={t("activity_form.billable")}
                checked={formik.values.billable}
                wrapperClassName={styles.billable}
                as={Checkbox}
              />
              <Field
                name="description"
                label={t("activity_form.description")}
                as={TextField}
                className={styles.description}
                isTextArea={true}
              >
                <FieldMessage
                  isError={
                    formik.errors.description && formik.touched.description
                  }
                  errorText={formik.errors.description}
                  alignRight={true}
                  hintText={`${formik.values.description.length} / 1024`}
                />
              </Field>
              <UploadImage
                activityId={props.activity?.id}
                imgBase64={imageBase64}
                handleChange={setImageBase64}
                hasImage={props.activity?.hasImage ?? false}
                toggleErrorModal={toggleErrorModal}
              />
            </div>
            <ActivityFormActions
              activity={props.activity}
              onRemove={handleRemoveActivity}
              onSave={formik.handleSubmit}
            />
          </form>
        )}
      </Formik>
      {errorModalIsOpen && (
        <ErrorModal
          message={{
            title: "Whoops!",
            description: "An error occurred"
          }}
          onClose={toggleErrorModal}
          onConfirm={toggleErrorModal}
          confirmText={t("actions.accept")}
        />
      )}
    </Fragment>
  );
};

export default ActivityForm;
