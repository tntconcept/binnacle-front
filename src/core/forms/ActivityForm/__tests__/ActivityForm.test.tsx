import React from "react"
import ActivityForm from "core/forms/ActivityForm/ActivityForm"
import {fireEvent, render, wait, waitForDomChange, waitForElement} from "@testing-library/react"
// @ts-ignore
import fetchMock from "fetch-mock/es5/client"
import {IProjectRole} from "interfaces/IProjectRole"
import {IActivity} from "interfaces/IActivity"
import {ORGANIZATION_ENDPOINT, PROJECT_ENDPOINT, PROJECT_ROLE_ENDPOINT} from "services/endpoints"

jest.mock('react-i18next', () => ({
  useTranslation: () => ({t: (key: string) => key})
}))

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

describe("ActivityForm", () => {

  const baseActivity: IActivity = {
    startDate: new Date(2020, 0, 2, 9, 0, 0),
    duration: 180,
    description: "Is Billable",
    billable: true,
    id: 1,
    userId: 200,
    organization: {
      id: 1,
      name: "Puma"
    },
    project: {
      id: 10,
      name: "Marketing",
      billable: true,
      open: true
    },
    projectRole: {
      id: 200,
      name: "Pixel perfect"
    }
  }

  beforeEach(() => {
    fetchMock
      .getOnce(`path:${ORGANIZATION_ENDPOINT}`, [
        {
          id: 1,
          name: "Adidas"
        },
        {
          id: 2,
          name: "Puma"
        }
      ])
      .getOnce(`end:${PROJECT_ENDPOINT}/1`, [
        {
          id: 10,
          name: "Marketing",
          open: true,
          billable: true
        },
        {
          id: 20,
          name: "User experience",
          open: true,
          billable: false
        }
      ])
      .getOnce(`end:${PROJECT_ROLE_ENDPOINT}/10`, [
        {
          id: 100,
          name: "Developer"
        },
        {
          id: 200,
          name: "Pixel perfect",
        }
      ])
  })
  afterEach(fetchMock.restore)

  describe("create a new activity", () => {
    it("should show the last activity role selected", function () {
      const lastSelectedRole: IProjectRole = {id: 15, name: "React developer"}

      const result = render(
        <ActivityForm
          activity={undefined}
          initialSelectedRole={lastSelectedRole}
          initialStartTime={undefined}
        />
      )

      expect(result.getByTestId("role_15")).toBeChecked()
    })

    it("should display select entities when the user makes his first-ever imputation", async () => {
      const result = render(
        <ActivityForm
          activity={undefined}
          initialSelectedRole={undefined}
          initialStartTime={undefined}
        />
      )

      await waitForDomChange()

      expect(
        result.getByText("activity_form.organization")
      ).toBeInTheDocument()
      expect(result.getByText("activity_form.project")).toBeInTheDocument()
      expect(result.getByText("activity_form.role")).toBeInTheDocument()
    })

    it("should create activity correctly", async () => {
      fetchMock.post('end:/api/activities', {foo: true})

      const form = render(<ActivityForm />)

      fireEvent.change(form.getByLabelText("activity_form.start_time"), {target: {value: '10:00'}})
      fireEvent.change(form.getByLabelText("activity_form.end_time"), {target: {value: '13:30'}})
      fireEvent.change(form.getByLabelText("activity_form.description"), {target: {value: "Lorem ipsum."}})

      // await waitForDomChange()

      fireEvent.change(
        form.getByTestId("organization_combobox"),
        {target: {value: "Adidas"}}
      )
      fireEvent.click(form.getByText("Adidas"))

      await waitForDomChange()

      fireEvent.change(
        form.getByTestId("project_combobox"),
        {target: {value: "Mark"}}
      )

      fireEvent.click(form.getByText("Marketing"))

      await waitForDomChange()

      fireEvent.change(
        form.getByTestId("role_combobox"),
        {target: {value: "Pixel"}}
      )
      fireEvent.click(form.getByText("Pixel perfect"))

      // await waitForDomChange()

      fireEvent.click(form.getByTestId("save_activity"))

      await wait()

      expect(JSON.parse(fetchMock.lastOptions().body)).toMatchSnapshot()
    })
  })

  it("should edit an activity", async () => {
    fetchMock.put('end:/api/activities', {foo: true})

    const form = render(<ActivityForm activity={baseActivity}/>)

    const newDescription = "Activity Test Description"

    fireEvent.change(form.getByLabelText("activity_form.description"), {target: {value: newDescription}})

    fireEvent.click(form.getByTestId("save_activity"))

    await wait()

    expect(JSON.parse(fetchMock.lastOptions().body).activity.description).toBe(newDescription)
  })

  it("should validate fields correctly", async () => {
    const result = render(<ActivityForm />)

    // set end time before start time (by default is 9:00)
    fireEvent.change(result.getByLabelText("activity_form.end_time"), {
      target: {value: "07:30"}
    })

    fireEvent.click(result.getByTestId("save_activity"))

    await waitForDomChange()

    expect(result.getByText("form_errors.end_time_greater")).toBeInTheDocument()

    fireEvent.change(result.getByLabelText("activity_form.start_time"), {
      target: {value: ""}
    })

    fireEvent.change(result.getByLabelText("activity_form.end_time"), {
      target: {value: ""}
    })

    fireEvent.click(result.getByTestId("save_activity"))

    await waitForDomChange()

    // One for each select
    expect(result.getAllByText("form_errors.select_an_option").length).toBe(3)

    // Start time, End time and Description
    expect(result.getAllByText("form_errors.field_required").length).toBe(3)

  })

  it("should delete the activity correctly", async () => {
    fetchMock.delete('end:/api/activities/1', {})

    const form = render(<ActivityForm activity={baseActivity} />)

    fireEvent.click(form.getByText('actions.remove'))

    const yesModalButton = await waitForElement(() => form.getByTestId('yes_modal_button'));

    fireEvent.click(yesModalButton)

    await wait()

    expect(fetchMock.lastCall()).toContainEqual("http://localhost:8080/api/activities/1")
  })

  it("should not delete the activity if the user selects no", async () => {
    const form = render(<ActivityForm activity={baseActivity} />)

    fireEvent.click(form.getByText('actions.remove'))

    const noModalButton = await waitForElement(() => form.getByTestId('no_modal_button'));

    fireEvent.click(noModalButton);

    await wait()

    expect(noModalButton).not.toBeInTheDocument()
    expect(form.getByText('actions.remove')).toBeInTheDocument()
  })

  it("should update billable selecting a project from frequent roles list", () => {
    // For now frequentRolesList is mocked
    console.warn("Removes this warning when frequentRolesList is implemented")

    // initialSelectedRole -> activity -> undefined
    const result = render(
      <ActivityForm
        activity={{
          ...baseActivity,
          billable: false,
          projectRole: {
            id: 15,
            name: "React developer"
          }
        }}
        initialStartTime={undefined}
        initialSelectedRole={undefined}
      />
    )

    expect(result.getByTestId("billable_no")).toBeChecked()
    fireEvent.click(result.getByTestId("role_10"), {target: {value: "NobodyKnows"}})

    expect(result.getByTestId("billable_yes")).toBeChecked()
  })


  it("should update billable selecting a project from select field", async () => {
    const result = render(
      <ActivityForm
        activity={undefined}
        initialSelectedRole={undefined}
        initialStartTime={undefined}
      />
    )

    // await waitForDomChange()

    fireEvent.change(
      result.getByTestId("organization_combobox"),
      {target: {value: "Adi"}}
    )

    fireEvent.click(result.getByText("Adidas"))

    await waitForDomChange()

    fireEvent.change(
      result.getByTestId("project_combobox"),
      {target: {value: "Marke"}}
    )

    fireEvent.click(result.getByText("Marketing"))

    await waitForDomChange()

    expect(result.getByTestId("billable_yes")).toBeChecked()
  })

  it("should display select entities filled with the activity's data when it's role has not been found in frequent roles list", async () => {
    const result = render(
      <ActivityForm
        activity={baseActivity}
        initialStartTime="09:00"
        initialSelectedRole={{
          id: 500,
          name: "Role used in another activity"
        }}
      />
    )

    // await waitForDomChange()

    expect(result.getByTestId("organization_combobox")).toHaveValue("Puma")
    expect(result.getByTestId("project_combobox")).toHaveValue("Marketing")
    expect(result.getByTestId("role_combobox")).toHaveValue("Pixel perfect")
  })

  describe.skip("Image section", () => {
    it("should upload an image", function () {
      throw Error("not implemented yet")
    })

    it("should download an image", function () {
      throw Error("not implemented yet")
    })
  })
})
