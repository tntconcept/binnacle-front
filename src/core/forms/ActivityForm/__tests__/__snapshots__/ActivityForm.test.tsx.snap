// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActivityForm create a new activity should create activity correctly 1`] = `
Object {
  "activity": Object {
    "billable": true,
    "description": "Lorem ipsum.",
    "duration": 210,
    "organization": Object {
      "id": 1,
      "name": "Adidas",
    },
    "project": Object {
      "billable": true,
      "id": 10,
      "name": "Marketing",
      "open": true,
    },
    "projectRole": Object {
      "id": 200,
      "name": "Pixel perfect",
    },
    "startDate": "2020-02-05T09:00:00.000Z",
    "userId": 0,
  },
}
`;
